File: T-Bank/Algorithms-and-Data-Structures-2025/4sprint-SolvingProblemMethodsPart1/contest/B-2D_суммы/solution.cpp
#include <iostream>
#include <vector>


void solution() {
    /*
    Time Complexity O(NM)
    Memory Complexity O(NM)
    */
    int n, m, k;
    std::scanf("%d %d %d", &n, &m, &k);
    std::vector<std::vector<long long>> prefixSum (n + 1, std::vector<long long>(m + 1, 0));
    for (int i = 0; i < n; ++i) {
        for (int j = 0; j < m; ++j) {
            long long num;
            std::scanf("%lld", &num);
            prefixSum[i + 1][j + 1] = prefixSum[i + 1][j] + prefixSum[i][j + 1] + num - prefixSum[i][j];
        }
    }
    for (int i = 0; i < k; ++i) {
        int y1, x1, y2, x2;
        std::scanf("%d %d %d %d", &y1, &x1, &y2, &x2);
        long long output = prefixSum[y2][x2];
        output -= prefixSum[y1 - 1][x2];
        output -= prefixSum[y2][x1 - 1];
        output += prefixSum[y1 - 1][x1 - 1];
        std::printf("%lld\n", output);
    }
}


int main() {
    solution();
}

File: T-Bank/Algorithms-and-Data-Structures-2025/4sprint-SolvingProblemMethodsPart1/contest/B-2D_суммы/solution.py
import sys


def solution() -> None:
    """
    Time Complexity O(NM)
    Memory Complexity O(NM)
    """
    n, m, k = map(int, sys.stdin.readline().rstrip().split())
    prefix_sum: list[list[int]] = [[0] * (m + 1) for _ in range(n + 1)]
    for i in range(n):
        nums: iter[int] = map(int, sys.stdin.readline().rstrip().split())
        for j in range(m):
            num: int = next(nums)
            prefix_sum[i + 1][j + 1] = prefix_sum[i][j + 1] + prefix_sum[i + 1][j] + num - prefix_sum[i][j]

    for _ in range(k):
        y1, x1, y2, x2 = map(int, sys.stdin.readline().rstrip().split())
        output: int = prefix_sum[y2][x2]
        output -= prefix_sum[y2][x1 - 1]
        output -= prefix_sum[y1 - 1][x2]
        output += prefix_sum[y1 - 1][x1 - 1]
        
        sys.stdout.write('{}\n'.format(output))


if __name__ == '__main__':
    solution()